packages/astro/src/assets/services/service.ts

```diff
@@ -245,62 +255,46 @@ export const baseService: Omit<LocalImageService, 'transform'> = {
 				.sort()
 				.map((density) => Math.round(targetWidth * density));
 
-			densityWidths.forEach((width, index) => {
-				const maxTargetWidth = Math.min(width, maxWidth);
-
-				// If the user passed dimensions, we don't want to add it to the srcset
-				const { width: transformWidth, height: transformHeight, ...rest } = options;
-
-				const srcSetValue = {
-					transform: {
-						...rest,
-					},
-					descriptor: `${densityValues[index]}x`,
-					attributes: {
-						type: `image/${targetFormat}`,
-					},
-				};
-
-				// Only set width and height if they are different from the original image, to avoid duplicated final images
-				if (maxTargetWidth !== imageWidth) {
-					srcSetValue.transform.width = maxTargetWidth;
-					srcSetValue.transform.height = Math.round(maxTargetWidth / aspectRatio);
-				}
-
-				if (targetFormat !== options.format) {
-					srcSetValue.transform.format = targetFormat;
-				}
-
-				srcSet.push(srcSetValue);
-			});
+			allWidths.push(
+				...densityWidths.map((width, index) => ({
+					width,
+					descriptor: `${densityValues[index]}x` as const,
+				}))
+			);
 		} else if (widths) {
-			widths.forEach((width) => {
-				const maxTargetWidth = Math.min(width, maxWidth);
-
-				const { width: transformWidth, height: transformHeight, ...rest } = options;
-
-				const srcSetValue = {
-					transform: {
-						...rest,
-					},
-					descriptor: `${width}w`,
-					attributes: {
-						type: `image/${targetFormat}`,
-					},
-				};
-
-				if (maxTargetWidth !== imageWidth) {
-					srcSetValue.transform.width = maxTargetWidth;
-					srcSetValue.transform.height = Math.round(maxTargetWidth / aspectRatio);
-				}
+			allWidths.push(...widths.map((width) => ({ width, descriptor: `${width}w` as const })));
+		}
 
-				if (targetFormat !== options.format) {
-					srcSetValue.transform.format = targetFormat;
+		// Caution: The logic below is a bit tricky, as we need to make sure we don't generate the same image multiple times
+		// When making changes, make sure to test with different combinations of local/remote images widths, densities, and dimensions etc.
```

Comment:
Added a "There be dragons" comment here because it's honestly super tricky to completely avoid making transforms that are functionally equivalent to another one, but don't have quite the exact same shape.

Common examples includes:
- A transform where `webp` is specified manually vs one where it's not. Due to the default, both of those results in the same image.
- A transform where `width` is specified, but it's equivalent to the width of the image, so it's not resized.